server/rest/dataset.py

```diff
@@ -535,51 +536,51 @@ def _getPostgresGeojsonData(self, item):
-                assetstore = self.model('assetstore').load(file['assetstoreId'])
-                adapter = assetstore_utilities.getAssetstoreAdapter(assetstore)
-                func = adapter.downloadFile(
-                    file, offset=0, headers=False, endByte=None,
-                    contentDisposition=None, extraParameters=None)
-            except Exception:
-                raise GirderException('Unable to load link target dataset.')
-            featureCollections = json.loads(''.join(list(func())))
-
-            valueLinks = sorted([x for x in geometryField['links']
-                                 if x['operator'] == '='])
-            constantLinks = [x for x in geometryField['links']
-                             if x['operator'] == 'constant']
-            mappedGeometries = {}
-            for feature in featureCollections['features']:
-                skip = False
-                for constantLink in constantLinks:
-                    if feature['properties'][constantLink['field']] != constantLink['value']:
-                        skip = True
-                        break
-                if skip:
-                    continue
-                try:
-                    key = ''.join([feature['properties'][x['field']] for x in valueLinks])
-                except KeyError:
-                    raise GirderException('missing property for key ' +
-                                          x['field'] + ' in geometry link target geojson')
-                mappedGeometries[key] = feature['geometry']
-
-            assembled = []
-            for record in records:
-                key = ''.join([record[x['value']] for x in valueLinks])
-                if key in mappedGeometries:
-                    assembled.append(
-                        geojson.Feature(geometry=mappedGeometries[key], properties=record)
-                    )
-            if len(assembled) == 0:
+            featureCollection, _ = self.linkAndAssembleGeometry(
+                geometryField['links'], geometryField['itemId'], records)
+            if len(featureCollection.features) == 0:
                 raise GirderException('Dataset is empty')
-
-            return geojson.FeatureCollection(assembled)
+            return featureCollection
+
+    def linkAndAssembleGeometry(self, link, linkItemId, records):
+        try:
+            item = self.model('item').load(linkItemId, force=True)
+            featureCollections = self.downloadDataset(item)
+        except Exception:
+            raise GirderException('Unable to load link target dataset.')
+
+        valueLinks = sorted([x for x in link
+                             if x['operator'] == '='])
+        constantLinks = [x for x in link
+                         if x['operator'] == 'constant']
+        mappedGeometries = {}
+        linkingDuplicateCount = 0
+        for feature in featureCollections['features']:
+            skip = False
+            for constantLink in constantLinks:
```

Comment:
@matthewma7 Thanks for the clarification. I would prefer a minor refactor here to make it easier to reason about.

Reply:
@dorukozturk Since the flow has been working fine (This logic has been used for the Postgres feature before), so I didn't change the logic flow but renamed variable and add a comment to make it easier to understand.